dm/pkg/binlog/pos_finder.go

```diff
@@ -0,0 +1,370 @@
+		position.Name = path.Join(r.relayDir, position.Name)
+		return binlogReader, binlogReader.StartSyncByPos(position)
+	}
+}
+
+func (r *binlogPosFinder) findMinTimestampOfBinlog(currBinlog binlogSize) (uint32, error) {
+	var minTs uint32
+	binlogReader, err := r.startSync(mysql.Position{Name: currBinlog.name, Pos: FileHeaderLen})
+	if err != nil {
+		return 0, err
+	}
+	for {
+		ev, err := binlogReader.GetEvent(r.tctx.Ctx)
+		if err != nil {
+			binlogReader.Close()
+			return 0, err
+		}
+		// break on first non-fake event(must be a format description event)
+		if !utils.IsFakeRotateEvent(ev.Header) {
+			minTs = ev.Header.Timestamp
+			break
+		}
+	}
+	binlogReader.Close()
+
+	return minTs, nil
+}
+
+func (r *binlogPosFinder) initTargetBinlogFile(ts int64) error {
+	targetTs := uint32(ts)
+	var lastTs, minTs uint32
+	var lastMid int
+	binaryLogs, err := r.getBinlogFiles()
+	if err != nil {
+		return err
+	}
+	if len(binaryLogs) <= 0 {
+		// should not happen on a master with binlog enabled
+		return errors.New("cannot find binlog files")
+	}
+
+	begin, end := 0, len(binaryLogs)-1
+	for begin <= end {
+		mid := (begin + end) / 2
+		currBinlog := binaryLogs[mid]
+
+		minTs, err = r.findMinTimestampOfBinlog(currBinlog)
+		if err != nil {
+			return err
+		}
+
+		r.tctx.L().Info("min timestamp in binlog file", zap.Reflect("file", currBinlog), zap.Uint32("ts", minTs))
+
+		lastTs = minTs
+		lastMid = mid
+
+		if minTs >= targetTs {
+			end = mid - 1
+		} else {
+			// current binlog maybe the target binlog file, we'll backtrace to it later.
+			begin = mid + 1
+		}
+	}
+	if lastTs >= targetTs {
+		if lastMid == 0 {
+			// timestamp of first binlog event in first binlog file >= targetTs
+			r.targetBinlog = binaryLogs[lastMid]
+			r.tsBeforeFirstBinlog = true
+		} else {
+			// timestamp of first event in lastMid >= targetTs, need to search from previous binlog file
+			r.targetBinlog = binaryLogs[lastMid-1]
+		}
+	} else {
+		r.targetBinlog = binaryLogs[lastMid]
+	}
+	r.lastBinlogFile = r.targetBinlog.name == binaryLogs[len(binaryLogs)-1].name
+	return nil
+}
+
+func (r *binlogPosFinder) processGTIDRelatedEvent(ev *replication.BinlogEvent, prevSet gtid.Set) (gtid.Set, error) {
+	ev, err := r.parser.Parse(ev.RawData)
+	if err != nil {
+		return nil, err
+	}
+	switch ev.Header.EventType {
+	case replication.PREVIOUS_GTIDS_EVENT:
+		newSet, err := event.GTIDsFromPreviousGTIDsEvent(ev)
+		if err != nil {
+			return nil, err
+		}
+		return newSet, nil
+	case replication.MARIADB_GTID_EVENT:
+		fallthrough
+	case replication.GTID_EVENT:
+		gtidStr, _ := event.GetGTIDStr(ev)
+		if err := prevSet.Update(gtidStr); err != nil {
+			return nil, err
+		}
+	case replication.MARIADB_GTID_LIST_EVENT:
+		newSet, err := event.GTIDsFromMariaDBGTIDListEvent(ev)
+		if err != nil {
+			return nil, err
+		}
+		return newSet, nil
+	}
+	return prevSet, nil
+}
+
+func (r *binlogPosFinder) checkTransactionBoundaryEvent(ev *replication.BinlogEvent) (bool, error) {
```

Comment:
```suggestion
func (r *binlogPosFinder) checkTransactionBeginEvent(ev *replication.BinlogEvent) (bool, error) {
```

more often used in codes of this PR